### ISSUE TITLE
Position accessors for docked text

### ISSUE THREAD

**Issue by gagern** [CindyJS/CindyJS#488]
Accessing the position of docked text must not throw an exception. This is a follow-up on #485.

In Cinderella, reading the position of a text which is docked to a point will read the position of the point without taking the text offset into account. Reading the position of text docked to the viewport boundary appears to just report the last position the text had while it was still free. Writing the position doesn't have any visible effect in either of these cases.

I think we can duplicate that behavior in CindyJS, but I'd prefer some more deterministic behavior for text which got docked to the viewport. I assume I'd want to return the user-space coordinates of the docking corner in this case. We could refactor [the “give me the position of that text element” code](https://github.com/CindyJS/CindyJS/blob/v0.8.0/src/js/libgeo/GeoRender.js#L254-L262) to a separate function, and then call that function both when rendering text and when reading the text position. When writing we'd add a check, and only perform the assignment if the text is not docked.

**Comment by kortenkamp** [comment]
Ah. Just wanted to file this.

Let me comment a bit: In Cinderella, docked elements report the position of the element they are docked to (not the last set, unless you set it and the docked-to element was not moved). And setting it is more or less undefined - the position will be reported back unless the docked-to element moved (I know I repeat what I said before). 

Here's a suggestion: setting xy of a docked element will change the offset in such a way that it is still docked, but now at position xy. Reading the position will report the current position in homogeneous coordinates. 

I also suggest to add a new accessor .dockedto that will either return the geometric element if it is docked to an element or something that will describe the ne, se, nw, sw corner. 

While doing this: Why not allow the same description that is used for reference points in images for docking points in viewports? That would be very, very nice – and I would be happy to implement that in Cinderella, if it makes it into CindyJS.

**Comment by gagern** [comment]
> Here's a suggestion: setting xy of a docked element will change the offset in such a way that it is still docked, but now at position xy. Reading the position will report the current position in homogeneous coordinates.

That's tricky because the position is in user coordinates while the offset is in canvas coordinates. Converting the offset into user coordinates as well would mean the position would change when you zoom, which I'd consider confusing. I'd prefer to keep the two concepts distinct and compatible with current Cinderella behavior. Having the offset accessible through an accessor would make sense, though. A `docketo` accessor to obtain the geometric element it's docked to sounds reasonable as well. Writing e.g. `false` to that might be used to un-dock text. I'm unsure how we would represent screen corners. Simply as text strings?

> Why not allow the same description that is used for reference points in images for docking points in viewports?

You mean with percentages? Sounds useful, yes. Is it `T.dockedto = ["50%","75%"]`, or is it `T.dockedto="viewport"; T.xy=["50%","75%"]`? I think the former makes more sense, since then you could still read (but not write) `T.xy` to obtain the user-space position of the text and do stuff with that. It would also mean that we could incorporate the screen corners described above by translating them to percentages.

**Comment by kortenkamp** [comment]
No, you misunderstood. I don't want to convert the offset to pixel coordinates (sorry, I feel more confident using pixel coordinates for what you call user space and homogeneous coordinates for what you call the other thing). I want to recalculate the offset (in pixels) in order to match the position (in homo). The same is happening for labels in Cinderella when you move them, or when you dock an element. 

That is, given A as the homo coord of the docked to element and (dx,dy) as offset, moving the docked element to B will calculate B-A in pixel coordinates and set (x,y) to this.

**Comment by kortenkamp** [comment]
For your second comment I would like to have a look at the current implementation first.

**Comment by gagern** [comment]
> I feel more confident using pixel coordinates for what you call user space and homogeneous coordinates for what you call the other thing

The other way round. User space is the space the geometry is expressed in. `P.homog`, `P.xy` and all that. Canvas space is what I'd have called device space a while back, except that with the advent of HiDPI or mobile devices, CSS pixels no longer correspond to device pixels. But it's pixels either way. The position is in user space (homog), the offset is in canvas space (pixels).
